### PR TITLE
exec: Implement detach option

### LINKED ISSUE #318: exec: Implement the detach option
`exec` command does not support the attached mode, by default, everything is run as detached.

### PR DESCRIPTION
**Pull request by sboeuf**
This patch implements the detach option for the exec command. In case
the flag is set to false, we stay attached to the process. This means
we have to wait for the end of the shim process and return the exit
code through the runtime itself.

Docker already implements this option on his own, meaning this patch
is useful for cases where the runtime is used as a standalone binary.

For instance, CRI-O uses the runtime with this option, meaning it
expects this to work as expected.

Fixes #318

### CODE REVIEW

File: exec.go [hunk 1]
```diff
@@ -210,5 +212,20 @@ func execute(params execParams) error {
...
 
+	if !params.detach {
+		p, err := os.FindProcess(process.Pid)
+		if err != nil {
+			return err
+		}
+
+		ps, err := p.Wait()
+		if err != nil {
+			return fmt.Errorf("Process state %s: %v", ps.String(), err)
```
Comment (jodh-intel) [inline exec.go]: It would be useful to include details of the container here.

File: exec.go [hunk 2]
```diff
@@ -210,5 +212,20 @@ func execute(params execParams) error {
 		return err
 	}
 
+	if !params.detach {
+		p, err := os.FindProcess(process.Pid)
+		if err != nil {
+			return err
+		}
+
+		ps, err := p.Wait()
```
Comment (jodh-intel) [inline exec.go]: This seems to suggest that `exec` is currently broken (fails to wait?)

**Comment by jodh-intel** [comment]
Can you add a test or two to `.ci/run.sh` so we have some level of test coverage for this change please?

File: exec.go (hunk 2 above)
Comment (sboeuf) [inline exec.go]: What do you mean ? No, exec is not broken, we just never implemented the behavior related to detach == false.

File: exec.go (hunk 1 above)
Comment (sboeuf) [inline exec.go]: I'll add that.

File: exec.go (hunk 2 above)
Comment (jodh-intel) [inline exec.go]: Right, currently all `exec`'s are "silently" detached but we never documented that (in the code or in an issue).

Comment (sboeuf) [inline exec.go]: Yes exactly, but it was not an issue from Docker perspective because Docker handle this on his own.

**Comment by jodh-intel** [comment]
Hi @sboeuf - any progress on this? We need to find a way to ensure this change is tested somehow.

**Comment by sboeuf** [comment]
Hey @jodh-intel, I have to work on the agent to implement missing things. I'll take care of this after that.

**Comment by jodh-intel** [comment]
Great - thanks @sboeuf!

**Comment by sboeuf** [comment]
@jodh-intel I have rebased the patch on the latest master, but I haven't added any new tests because this has to be done on https://github.com/clearcontainers/tests as mentioned in the issue that I have just opened https://github.com/clearcontainers/tests/issues/171.
Indeed, the standalone tests for the runtime are run with ginkgo. We don't want to duplicate on this repo.

**Comment by sboeuf** [comment]
@dlespiau please take a look.

**Comment by devimc** [comment]
lgtm

**Comment by sboeuf** [comment]
Here is are the tests https://github.com/clearcontainers/tests/pull/180

**Pull request merged**